pandas/core/frame.py

```diff
@@ -9787,7 +9787,9 @@ def _join_compat(
                     "Suffixes not supported when joining multiple DataFrames"
                 )
 
-            frames = [self] + list(other)
+            frames = [cast("DataFrame | Series", self)] + list(
```

Comment:
Since this is in `frame.py`, I don't think you need to cast `self` because `self` is a `DataFrame`.  Also, add a comment indicating why the second cast is needed - it's because the `mypy` can't infer that `other` is a `list` at this point.

Reply:
I believe mypy doesn't like adding together two lists of different types.

Reply:
Yes, I think you're right.